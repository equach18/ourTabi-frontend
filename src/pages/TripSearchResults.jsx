import { useEffect, useState } from "react";
import { useSearchParams } from "react-router-dom";
import OurTabiApi from "../api/ourTabiApi";
import TripCard from "../components/trip/TripCard";

function TripSearchResults() {
  const [searchParams] = useSearchParams();
  const term = searchParams.get("term") || "";
  const [results, setResults] = useState([]);

  useEffect(() => {
    async function fetchResults() {
      try {
        const trips = await OurTabiApi.getPublicTrips({ title: term });
        setResults(trips);
      } catch (err) {
        console.error("Search error:", err);
      }
    }
    if (term) fetchResults();
  }, [term]);

  return (
    <div className="h-[calc(100vh-4rem)] px-50 py-12">
      <h2 className="text-xl font-semibold mb-4">
        Search results for: <span className="text-blue-600">{term}</span>
      </h2>
      {results.length > 0 ? (
        results.map((trip) => (
          <div key={trip.id} className="mb-4">
            <TripCard trip={trip} />
          </div>
        ))
      ) : (
        <p className="text-gray-500">No trips found.</p>
      )}
    </div>
  );
}

export default TripSearchResults;
