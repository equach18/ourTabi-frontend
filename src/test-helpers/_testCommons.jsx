import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserContext from "../context/UserContext";
import TripContext from "../context/TripContext";

export const testUser1 = {
  id: 1,
  firstName: "Test1",
  lastName: "User1",
  username: "testUser1",
  profilePic: "",
};
export const testUser2 = {
  id: 2,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser2",
  profilePic: "testUser2.png",
};
export const testUser3 = {
  id: 3,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser3",
  profilePic: "testUser3.png",
};

export const defaultMembers = [
  {
    userId: testUser1.id,
    username: testUser1.username,
    profilePic: testUser1.profilePic,
  },
  {
    userId: testUser2.id,
    username: testUser2.username,
    profilePic: testUser2.profilePic,
  },
];

export const testActivity = {
  id: 1,
  name: "Kayaking",
  tripId: 53,
  category: "Other",
  location: "Chicago",
  description: "So much fun",
  scheduledTime: "2025-07-10T14:30:00Z",
};

export const testComment = {
  id: 1,
  text: "Test comment",
  createdAt: "2025-07-11T14:30:00Z",
  userId: 1,
};

export const testTrip = {
  id: 42,
  title: "Test trip",
  destination: "Chicago, IL",
  startDate: "2025-07-10",
  endDate: "2025-07-12",
  isPrivate: true,
};

export const testFriendsData = {
  friends: [testUser1 ],
  incomingRequests: [testUser2],
  sentRequests: [testUser3],
};

// Default mock context values
const defaultUserContext = { currentUser: testUser1 };
const defaultTripContext = {
  voteOnActivity: vi.fn(),
  votes: {},
  deleteComment: vi.fn(),
  addMember: vi.fn(),
  removeMember: vi.fn(),
  members: defaultMembers,
};

// create mock trip context
export function createMockTripContext(overrides = {}) {
  return {
    voteOnActivity: vi.fn(),
    votes: {},
    deleteComment: vi.fn(),
    members: [],
    addMember: vi.fn(),
    removeMember: vi.fn(),
    ...overrides,
  };
}

/**
 * Renders a component inside MemoryRouter
 */
export function renderWithRouter(ui) {
  return render(<MemoryRouter>{ui}</MemoryRouter>);
}

/**
 * Renders a component inside MemoryRouter + UserContext + TripContext
 */
export function renderWithAllProviders(
  ui,
  {
    userContext = defaultUserContext,
    tripContext = defaultTripContext,
    route = "/",
  } = {}
) {
  return render(
    <MemoryRouter initialEntries={[route]}>
      <UserContext.Provider value={userContext}>
        <TripContext.Provider value={tripContext}>{ui}</TripContext.Provider>
      </UserContext.Provider>
    </MemoryRouter>
  );
}

/**
 * Renders a component with just UserContext
 */
export function renderWithUserProvider(ui, { userContext } = {}) {
  return render(
    <UserContext.Provider value={userContext}>{ui}</UserContext.Provider>
  );
}
