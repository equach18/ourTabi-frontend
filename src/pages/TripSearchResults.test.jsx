import { screen, waitFor } from "@testing-library/react";
import { Route, Routes } from "react-router-dom";
import TripSearchResults from "./TripSearchResults";
import { renderWithAllProviders } from "../test-helpers/_testCommons";

vi.mock("../components/trip/TripCard", () => ({
  default: ({ trip }) => <div data-testid="TripCard">{trip.title}</div>,
}));

vi.mock("../api/ourTabiApi", () => ({
  default: {
    getPublicTrips: vi.fn(),
  },
}));
import OurTabiApi from "../api/ourTabiApi";

function renderSearchResults(term = "") {
  return renderWithAllProviders(
    <Routes>
      <Route path="/search" element={<TripSearchResults />} />
    </Routes>,
    { route: `/search?term=${term}` }
  );
}

describe("TripSearchResults", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("matches snapshot when there is no search term", () => {
    const { asFragment } = renderSearchResults("");
    expect(asFragment()).toMatchSnapshot();
  });
  it("displays msg when no trips are found", async () => {
    OurTabiApi.getPublicTrips.mockResolvedValueOnce([]);

    renderSearchResults("sdfdsfjk");

    await waitFor(() =>
      expect(screen.getByText(/No trips found/i)).toBeInTheDocument()
    );
  });

  it("displays trips when results are returned", async () => {
    OurTabiApi.getPublicTrips.mockResolvedValueOnce([
      { id: 1, title: "Paris Getaway" },
      { id: 2, title: "London Trip" },
    ]);

    renderSearchResults("europe");

    await waitFor(() =>
      expect(screen.getAllByTestId("TripCard")).toHaveLength(2)
    );
    expect(screen.getByText("Paris Getaway")).toBeInTheDocument();
    expect(screen.getByText("London Trip")).toBeInTheDocument();
  });

  it("shows the search term in the heading", async () => {
    OurTabiApi.getPublicTrips.mockResolvedValueOnce([]);

    renderSearchResults("japan");

    await waitFor(() =>
      expect(
        screen.getByText(
          (content, node) => node.textContent === "Search results for: japan"
        )
      ).toBeInTheDocument()
    );
  });

  it("calls the API with the correct search term", async () => {
    OurTabiApi.getPublicTrips.mockResolvedValueOnce([]);

    renderSearchResults("beach");

    await waitFor(() => {
      expect(OurTabiApi.getPublicTrips).toHaveBeenCalledWith({
        title: "beach",
      });
    });
  });

  it("handles API errors gracefully", async () => {
    console.error = vi.fn(); // suppress expected error logging
    OurTabiApi.getPublicTrips.mockRejectedValueOnce(new Error("Network error"));

    renderSearchResults("error");

    await waitFor(() => {
      expect(screen.getByText(/No trips found/i)).toBeInTheDocument();
    });
  });
});
