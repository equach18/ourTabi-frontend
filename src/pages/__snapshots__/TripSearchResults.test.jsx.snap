// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TripSearchResults > matches snapshot when there is no search term 1`] = `
<DocumentFragment>
  <div
    class="h-[calc(100vh-4rem)] px-50 py-12"
  >
    <h2
      class="text-xl font-semibold mb-4"
    >
      Search results for: 
      <span
        class="text-blue-600"
      />
    </h2>
    <p
      class="text-gray-500"
    >
      No trips found.
    </p>
  </div>
</DocumentFragment>
`;
